drivers/sht11: Major refactoring

maribu opened the pull request:
- Renamed driver to sht1x to reflect that fact SHT10, SHT11 and SHT15 are supported
- Removed sensor specific shell commands in favor of SAUL
- Major refactoring and new features:
    - Use RIOT's GPIO interface to access the sensor to increase portability
    - Changed API to allow more than one sensor per board
    - Added `sht1x_params.h` that specifies how the sensors is connected - each board can overwrite default settings by #defining SHT1X_PARAM_CLK and SHT1X_PARAM_DATA
    - Changed arithmetic to use integer calculations only instead of floating point arithmetic
    - Added support for checking the CRC sum
    - Allow optional skipping of the CRC check to speed up measuring
    - Added support for advanced features like reducing the resolution and skipping calibration to speed up measuring
    - Allow specifying the supply voltage of sensor which heavily influences the temperature result (and use that information to calculate the correct temperature)
    - Reset sensor on initialization to bring it in a well known state
    - Support for the obscure heater feature. (Can be useful to check the temperature sensor?)
- Confirmed new integer only arithmetic by unit tests
- Added SAUL integration

Hint: Before testing on the MSBA2 or other lpc2387 based boards, please apply [this PR](https://github.com/RIOT-OS/RIOT/pull/9296) containing various bugfixes for the gpio driver.

aabadie commented on drivers/sht1x/sht1x.c:
```diff
@@ -0,0 +1,665 @@
+        if (retval != 0)
+            return retval;
+
+        switch (write_byte(dev, SHT1X_STATUS_REG_W)){
```
Maybe this file could be uncrustified ? (a space is missing before `{` here in a lot of other places).

aabadie commented on drivers/sht1x/sht1x_saul.c:
```diff
@@ -0,0 +1,157 @@
+    .type = SAUL_SENSE_HUM
+};
+
+const saul_driver_t sht1x_saul_conf_driver = {
```
Why do you want to map the device configuration with SAUL ?

aabadie commented on sys/auto_init/auto_init.c:
```diff
@@ -20,8 +20,8 @@
 
-#ifdef MODULE_SHT11
-#include "sht11.h"
+#ifdef MODULE_SHT1X
```
This can be removed in fact.

aabadie commented on sys/auto_init/saul/auto_init_sht1x.c:
```diff
@@ -0,0 +1,104 @@
+        saul_entries[(i * 3) + 2].dev = &(sht1x_devs[i]);
+        saul_entries[(i * 3)    ].name = "SHT1X temperature";
+        saul_entries[(i * 3) + 1].name = "SHT1X humidity";
+        saul_entries[(i * 3) + 2].name = "SHT1X config (Low resolution | "
```
I think the calibration should be passed in the driver params.

maribu replied:
This was not actually calibrating the sensor - the calibration data is stored inside the sensor. But the sensor can be configured to not use them, which speeds up measurements about 10 ms. So this should be a runtime option, as this trade-off between speed and precision depends on the situation.

But the configuration was removed from SAUL anyway (as suggested), so that comment no longer applies.

aabadie commented on sys/shell/commands/Makefile:
```diff
@@ -8,9 +8,6 @@ endif
 ifneq (,$(filter ps,$(USEMODULE)))
   SRC += sc_ps.c
 endif
-ifneq (,$(filter sht11,$(USEMODULE)))
```
I would keep the shell commands

aabadie commented on tests/driver_sht1x/Makefile:
```diff
@@ -0,0 +1,6 @@
+BOARD ?= native
+include ../Makefile.tests_common
+
+USEMODULE += sht1x
```
Same comment about the version of sensor, better use the pseudomodule which will implicitly load the common concrete module.

maribu replied:
Sure the device specific pseudo modules should be used for unit tests that are not device specific? I would prefer to keep sht1x here, otherwise it is not obvious that is applies to all variants.

aabadie commented on tests/driver_sht1x/main.c:
```diff
@@ -0,0 +1,140 @@
+                    continue;
+                }
+                int16_t got_hum = sht1x_humidity(&dev, raw_hum, temp);
+                if (
```
This file needs to be uncrustified

aabadie commented on tests/driver_sht1x/main.c:
```diff
@@ -0,0 +1,140 @@
+    const char *res_temp[] = {"12", "14"};
+    const char *res_hum[] = {"8", "12"};
+    const char *voltages[] = {"5.0", "4.0", "3.5", "3.0", "2.5"};
+    sht1x_dev_t dev = {.conf = 0};
```
Why not use the `sht1x_init` with the default params ? This is the common usage for other sensors.

maribu replied:
Because this would require an actual sensor to be present, but the unit tests are intended to run on "native" without any hardware. The unit tests verify that the integer arithmetic still gives the same results as the floating point arithmetic given in the data sheet. Because these tests are computationally expensive, it makes no sense to run them on embedded hardware.

Also, using default parameters would dramatically decrease test coverage. The unit tests iterate over all possible configuration states in the for loops and verify that the calculations are correct for all of them - including the configuration that was chosen to be the default one.

maribu commented on drivers/include/sht1x.h:
```diff
@@ -0,0 +1,187 @@
+ * @file
+ * @brief       SHT1X Device Driver
+ *
+ * @author      Freie Universität Berlin, Computer Systems & Telematics
```
Yes, I could remove the author tag and keep them in the copyright declaration. The author tag does not provide additional information compared to the copyright declaration.

maribu commented:
@aabadie: Thanks for your review. Unless I have overlooked anything, I have addressed all issues you pointed out with the exception of the issues for the unit tests I commented on.

Please review again and give feedback if the two points with the unit tests can remain as they are for the reasons I stated in the comments above, or whether I should address them somehow.

Kind regards,
Marian

aabadie commented on sys/auto_init/saul/auto_init_sht1x_saul.c:
```diff
@@ -0,0 +1,71 @@
+extern const saul_driver_t sht1x_saul_hum_driver;
+/** @} */
+
+void auto_init_sht1x_saul(void)
```
No need for a saul specific function, you could add it to `auto_init_sht1x` but guard it with an `#ifdef` if the saul module is loaded. If you do this, `auto_init_sht1x_saul.c` can be removed. And the `auto_init_sht1x.c` could be with the other saul devices auto init files.

maribu commented on sys/shell/commands/sc_sht1x.c:
```diff
@@ -0,0 +1,382 @@
+    const sht1x_dev_t *dev = get_dev(argc, argv);
+
+    if (!dev) {
+        return -1;
```
The error message is printed by `get_dev()` :-)

maribu commented:
@aabadie: Thanks for the review and for pointing me to the unit tests app. This is much better now :-)

aabadie commented on drivers/include/sht1x.h:
```diff
@@ -0,0 +1,186 @@
+ *
+ * @param dev         Device from which the raw value was received
+ * @param raw         The raw (unprocessed) temperature value
+ * @param temp        The temperate at which the humidity was measure in E-02 °C
```
typo 'temperature'

aabadie commented on sys/auto_init/auto_init.c:
```diff
@@ -20,8 +20,8 @@
-#ifdef MODULE_SHT11
-#include "sht11.h"
+#ifdef MODULE_SHT1X
+void auto_init_sht1x(void);
```
No need to declare this function here. Just use `extern` before the call to the function.

aabadie commented on sys/auto_init/auto_init.c:
```diff
@@ -101,9 +101,9 @@ void auto_init(void)
-#ifdef MODULE_SHT11
-    DEBUG("Auto init SHT11 module.\n");
-    sht11_init();
+#ifdef MODULE_SHT1X
```
I would move this just before the `#ifdef MODULE_AUTO_INIT_SAUL`.

maribu commented:
Unless I have overlooked anything, I have address all open issues and also added an on-device test for the driver. This test will read both temperature and humidity value three times for each configuration and then drop to the shell for manual testing. I added the MSBA2 as default board, because I tested on that. Please note that the driver does only work on the MSBA2 after the gpio driver bugs addressed by [this PR](https://github.com/RIOT-OS/RIOT/pull/9296) are fixed.

aabadie commented on tests/driver_sht1x/main.c:
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2018 2018 Otto-von-Guericke-Universität Magdeburg
```
2018 appears twice

maribu commented on drivers/sht1x/sht1x.c:
```diff
@@ -0,0 +1,694 @@
+        return retval;
+    }
+
+    {
```
This limits the validity of the variables `crc` and `expected` to that scope.

maribu commented on drivers/sht11/sht11.c:
```diff
@@ -1,350 +0,0 @@
-/*
```
When looking at the [specific commit](https://github.com/RIOT-OS/RIOT/pull/9317/commits/c0903e509602395604635e5895020bf4c57e8dc1) the rename seemed to work, but in the overview of all changes it still looks like one file removed and another created

maribu commented:
In case you wonder why I put thumbs up on every comment: I in fact missed some and would like to tick them with a check mark. As there is no check mark, the thumbs  up is what comes nearest to it. So it is just for me to check that I actually applied every comment

maribu commented:
Done :-)

aabadie commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -105,6 +106,10 @@ PSEUDOMODULES += si7013
 PSEUDOMODULES += si7020
 PSEUDOMODULES += si7021
 
+# include variants of SX127X drivers as pseudo modules
```
Ok, I understand. This block should be moved to its original place (before the si1145 pseudo modules). Normally only changes related to sht11/10/15 should appear.

aabadie commented on drivers/sht1x/include/sht1x_defines.h:
```diff
@@ -0,0 +1,72 @@
+/** @} */
+
+/**
+ * @name Commands that can be send to the SHT1X driver
```
s/send/sen**t**/

aabadie commented:
Just triggered Murdock, maybe some other things will raise.

maribu commented:
Done. I believe I still have to add some boards to BOARD_INSUFFICIENT_MEMORY for the test, but lets wait for Murdock to complete

aabadie commented:
Murdock is still failing, I think you have to remove the `DIRS = drivers` in `boards/common/msb-430/Makefile`

maribu commented:
The default pin configuration is invalid on telosb, wsn430-v1_3b, wsn430-v1_4 and z1. How should I handle that?

aabadie commented:
> The default pin configuration is invalid on telosb, wsn430-v1_3b, wsn430-v1_4 and z1. How should I handle that?

No idea, this is related to 16bit platform but the error is weird. Maybe @cladmi has an idea ?

maribu commented:
Hi,

`cpu/msp430fxyz/include/periph_cpu.h` contains:
``` C
typedef uint16_t gpio_t;
[...]
#define GPIO_PIN(x, y)      ((gpio_t)(((x & 0xff) << 8) | (1 << (y & 0xff))))
```

so for `y > 15` the result of `(1 << (y & 0xff))` is (as far as I know) not strictly defined, as `a << b` is only defined for `0 <= b < BITWIDTH(a)`. Changing that line to 

```
#define GPIO_PIN(x, y)      ((gpio_t)(((x & 0xff) << 8) | (1 << (y & 0x0f))))
```

would make the result valid again. But maybe it is desired that the compilation fails for invalid pin specifications?

PS: I'm checking that right now with a dummy commit added on top of the PR. So please don't merge until I remove that again.

maribu commented:
@aabadie: I just rebased the PR against current master (and removed the testing commit). The issue with compilation on msp430fxyz boards [is now fixed](https://github.com/RIOT-OS/RIOT/pull/9367) :-)
